pypose/module/LQR.py

```diff
@@ -0,0 +1,235 @@
+import torch as torch
+import time
+import torch.nn as nn
+
+
+class DP_LQR(nn.Module):
+    r'''
+    Discrete-time finite-horizon Linear Quadratic Regulator (LQR)
+
+    Args:
+        n_state (int): Parameter that determines the dimension of the state.
+        n_ctrl (int): Parameter that determines the dimension of the input.
+        T (int): Total time steps.
+        system (instance): The system sent to LQR to affine state-transition dynamics.
+        A (:obj:`Tensor`): The state matrix of LTI system.
+        B (:obj:`Tensor`): The input matrix of LTI system.
+        C (:obj:`Tensor`): The output matrix of LTI system.
+        D (:obj:`Tensor`): The observation matrix of LTI system,
+        c1 (:obj:`Tensor`): The constant input of LTI system,
+        c2 (:obj:`Tensor`): The constant output of LTI system.
+
+    LQR finds the optimal nominal trajectory :math:`\mathbf{\tau}_{1:T}^*` = 
+    :math:`\begin{Bmatrix} \mathbf{x}_t, \mathbf{u}_t \end{Bmatrix}_{1:T}` 
+    by solving the optimization problem
+
+    .. math::
+        \begin{aligned}
+            \mathbf{\tau}_{1:T}^* = \mathop{\arg\min}\limits_{\tau_{1:T}} \sum\limits_t\frac{1}{2}
+            \mathbf{\tau}_t^\top\mathbf{Q}_t\mathbf{\tau}_t + \mathbf{p}_t^\top\mathbf{\tau}_t \\
+            \mathrm{s.t.} \quad \mathbf{x}_1 = \mathbf{x}_{init}, 
+            \ \mathbf{x}_{t+1} = \mathbf{F}_t\mathbf{\tau}_t + \mathbf{f}_t \\
+        \end{aligned}
+
+    where :math:`\mathbf{\tau}` = :math:`\begin{bmatrix} \mathbf{x} & \mathbf{u} \end{bmatrix}^\top`, 
+    :math:`\mathbf{F}` = :math:`\begin{bmatrix} \mathbf{A} & \mathbf{B} \end{bmatrix}^\top`, 
+    :math:`\mathbf{f}` = :math:`\mathbf{c}_1`.
+
+    From a policy learning perspective, this can be interpreted as a module with unknown parameters
+    :math:`\begin{Bmatrix} \mathbf{Q}, \mathbf{p}, \mathbf{F}, \mathbf{f} \end{Bmatrix}`, 
+    which can be integrated into a larger end-to-end learning system.
+
+    Note:
+        Here, we consider the system sent to LQR as an LTI system in each small horizon. 
+        While it is still available for LTV systems.
```

Comment:
Here, we consider the system sent to LQR as a linear system in each small horizon. (We don't need to separate into LTI/LTV because it should work for general linear system)